text/0000-io-fs-2.1.md

```diff
@@ -0,0 +1,576 @@
+    pub trait CreateDirOptionsExt {
+        fn security_attributes(&mut self, ...) -> &mut Self;
+    }
+    impl CreateDirOptionsExt for CreateDirOptions { ... }
+}
+```
+
+This sort of builder is also extendable to other flavors of functions in the
+future, such as [C++'s template parameter][cpp-dir-template]:
+
+[cpp-dir-template]: http://en.cppreference.com/w/cpp/experimental/fs/create_directory
+
+```rust
+/// Use the specified directory as a "template" for permissions and security
+/// settings of the new directories to be created.
+///
+/// On unix this will issue a `stat` of the specified directory and new
+/// directories will be created with the same permission bits. On Windows
+/// this will trigger the use of the `CreateDirectoryEx` function.
+pub fn template<P: AsRef<Path>>(&mut self, path: P) -> &mut Self;
+```
+
+At this time, however, it it not proposed to add this method to
+`CreateDirOptions`.
+
+## Adding `FileType`
+
+Currently there is no enumeration or newtype representing a list of "file types"
+on the local filesystem. This is partly done because the need is not so high
+right now. Some situations, however, imply that it is more efficient to learn
+the file type at once instead of testing for each individual file type itself.
+
+For example some platforms' `DirEntry` type can know the `FileType` without an
+extra syscall. If code were to test a `DirEntry` separately for whether it's a
+file or a directory, it may issue more syscalls necessary than if it instead
+learned the type and then tested that if it was a file or directory.
+
+The full set of file types, however, is not always known nor portable across
+platforms, so this RFC proposes the following hierarchy:
+
+```rust
+#[derive(Copy, Clone, PartialEq, Eq, Hash)]
+pub struct FileType(..);
+
+impl FileType {
+    pub fn is_dir(&self) -> bool;
+    pub fn is_file(&self) -> bool;
+    pub fn is_soft_link(&self) -> bool;
+}
+```
+
+Extension traits can be added in the future for testing for other more flavorful
+kinds of files on various platforms (such as unix sockets on unix platforms).
+
+#### Dealing with `is_{file,dir}` and `file_type` methods
+
+Currently the `fs::Metadata` structure exposes stable `is_file` and `is_dir`
+accessors. The struct will also grow a `file_type` accessor for this newtype
+struct being added. It is proposed that `Metadata` will retain the
+`is_{file,dir}` convenience methods, but no other "file type testers" will be
+added.
+
+## Adding `fs::equivalent`
+
+A new function `equivalent` will be added to the `fs` module along the lines of
+[C++'s equivalent function][cpp-equivalent]:
+
+[cpp-equivalent]: http://en.cppreference.com/w/cpp/experimental/fs/equivalent
+
+```rust
+/// Test whether the two paths provided are equivalent references to the same
+/// file or directory.
+///
+/// This function will ensure that the two paths have the same status and refer
+/// to the same file system entity (e.g. at the same phyical location).
```

Comment:
I believe this function would surely return truthy value if one path is a symlink of another. It would return truthy value on hard link as well, because inode (on unix) is still the same.

More modern filesystems, such as btrfs, have a CoW feature implemented. This is done with a new type of links called reflinks. In short, a `cp --reflink from to` will create a new inode for file `to`, but pointing at the same location in the physical medium as `from` inode does. `stat` provides absolutely no information about the files being at the same physical location and I am not aware of a method to discern whether the file is a reflink or not without being the disk driver or invoking a copy on write (resulting in a constant false; a big no no).

So, this documentation as written (if we interpret the “e.g.” part as being normative) forces us to do the impossible or is too unclear to make the function useful.